pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go

```diff
@@ -0,0 +1,436 @@
+			}
+			if oldValue == nil || currDeviation.GreaterThan(field.Deviation) {
+				shouldReport = true
+			}
+		}
+
+		(*currentState)[field.InputKey] = singleValue
+		if len(field.OutputKey) > 0 {
+			report[field.OutputKey] = singleValue
+		} else {
+			report[field.InputKey] = singleValue
+		}
+	}
+
+	// If no AggregationField defines deviation, always report
+	hasNoDeviation := true
+	for _, field := range a.config.Fields {
+		if field.DeviationType != "none" {
+			hasNoDeviation = false
+			break
+		}
+	}
+	if hasNoDeviation {
+		shouldReport = true
+	}
+
+	stateValuesMap, err := values.WrapMap(currentState)
+	if err != nil {
+		return nil, err
+	}
+	stateBytes, err := proto.Marshal(values.ProtoMap(stateValuesMap))
+	if err != nil {
+		return nil, err
+	}
+
+	toWrap, err := formatReport(report, a.config.ReportFormat)
+	if err != nil {
+		return nil, err
+	}
+	reportValuesMap, err := values.NewMap(map[string]any{
+		a.config.OutputFieldName: toWrap,
+	})
+	if err != nil {
+		return nil, err
+	}
+	reportProtoMap := values.Proto(reportValuesMap).GetMapValue()
+
+	lggr.Debugw("Aggregation complete", "shouldReport", shouldReport)
+
+	return &types.AggregationOutcome{
+		EncodableOutcome: reportProtoMap,
+		Metadata:         stateBytes,
+		ShouldReport:     shouldReport,
+	}, nil
+}
+
+func (a *reduceAggregator) initializeCurrentState(lggr logger.Logger, previousOutcome *types.AggregationOutcome) (*map[string]values.Value, error) {
+	currentState := map[string]values.Value{}
+
+	if previousOutcome != nil {
+		pb := &pb.Map{}
+		proto.Unmarshal(previousOutcome.Metadata, pb)
+		mv, err := values.FromMapValueProto(pb)
+		if err != nil {
+			return nil, err
+		}
+		err = mv.UnwrapTo(currentState)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	zeroValue, _ := values.Wrap(decimal.NewFromInt(0))
+	for _, field := range a.config.Fields {
+		if _, ok := currentState[field.InputKey]; !ok {
+			currentState[field.InputKey] = zeroValue
+			lggr.Debugw("initializing empty onchain state for feed", "fieldInputKey", field.InputKey)
+		}
+	}
+
+	lggr.Debugw("current state initialized", "state", currentState, "previousOutcome", previousOutcome)
+	return &currentState, nil
+}
+
+func (a *reduceAggregator) extractValues(lggr logger.Logger, observations map[ocrcommon.OracleID][]values.Value, aggregationKey string) (vals []values.Value) {
+	for nodeID, nodeObservations := range observations {
+		// we only expect a single observation per node
+		if len(nodeObservations) == 0 || nodeObservations[0] == nil {
+			lggr.Warnf("node %d contributed with empty observations", nodeID)
+			continue
+		}
+		if len(nodeObservations) > 1 {
+			lggr.Warnf("node %d contributed with more than one observation", nodeID)
+			continue
+		}
+
+		val, err := nodeObservations[0].Unwrap()
+		if err != nil {
+			lggr.Warnf("node %d contributed a Value that could not be unwrapped", nodeID)
+			continue
+		}
+
+		// extract the value from complex types and re-wrap them
```

Comment:
A better comment here would explain why we need to do this: i.e. what feature of the value/aggregation is making it necessary to do this :)

Reply:
Expanding this comment makes me realize that I could just pass `any` type around. Keeping it as a Value doesn't really give any benefit. Where it's used for mode to get the protobuf I could just wrap it there.

Reply:
May not be a problem to move it around as a value though :grin:

Reply:
Yeah that makes sense :)